man/im_demand_hint.3
.TH IM_IOCHECK 3 "11 April 1990"
.SH NAME
im_demand_hint \- hint on demand style for im_generate(3)
.SH SYNOPSIS
#include <vips/vips.h>

int im_demand_hint( im, hint, in1, in2, ..., NULL )
.br
IMAGE *im, *in1, *in2, ...;
.br
im_demand_type hint;

int im_demand_hint_array( im, hint, in )
.br
IMAGE *im, **in;
.br
im_demand_type hint;
.SH DESCRIPTION
.B im_demand_hint(3) 
suggests to im_generate(3) the sorts of demand with which this image processing
operation would be happiest. 

.B im
is the image this operation is generating.
.B hint
is the demand style this operation would like (see below), and
.B in1 ...
is a NULL-terminated list of the image upon which this output image directly
depends, that is, the images which this operation will call im_prepare(3) for.

This list of parent images is necessary, as im_demand_hint(3) needs to know
what demand style this operation's ancestors have requested. If an ancestor of
this operation has specified a very restrictive demand style, then this
operation must fall back to that restrictive style and ignore the hint given
in this call to im_demand_hint(3).

VIPS currently supports three demand styles. More may be added in the future.
These demand styles are given below in order of increasing restrictiveness. 
When demanding output from a pipeline, im_generate(3) will use the most
restrictive of the styles requested by the operations in the pipeline.

IM_THINSTRIP
.br
This operation would like to output strips the width of the image and a few
pels high. This is option suitable for point-to-point operations, such as
those in the arithmetic package.

This option is only efficient for cases where each output pel depends upon the
pel in the corresponding position in the input image.

IM_FATSTRIP
.br
This operation would like to output strips the width of the image and as high
as possible. This option is suitable for area operations which do not
violently transform coordinates, such as im_conv(3). 

IM_SMALLTILE
.br
This is the most general demand format, and is the default. Output is
demanded in small (around 100x100 pel) sections. This style works reasonably
efficiently, even for bizzare operations like 45 degree rotate.

IM_ANY
.br
This image is not being demand-read from a disc file (even indirectly) so any
demand style is OK. It's used for things like
.B im_black(3)
where the pixels are calculated.

.B
im_demand_hint_array(3)
works exactly as im_demand_hint(3), but expects a pointer to a NULL-terminated
array of parent images as its third argument. You may use
im_allocate_input_array(3), if you wish, to build this structure.

As an example, here is part of the code for im_invert(3). In this operation,
each output pel depends upon the corresponding input pel. In other words,
there is no coordinate transformation in im_prepare(3). This style of
operation is most efficient with IM_THINSTRIP IO.

int im_invert( IMAGE *in, IMAGE *out )
.br
{
.br
	if( in->Coding != NOCODING ) {
.br
		im_errormsg( "im_invert: input coded" );
.br
		return( -1 );
.br
	}
.br
	if( in->BandFmt != FMTUCHAR ) {
.br
		im_errormsg( "im_invert: input not UCHAR" );
.br
		return( -1 );
.br
	}
.br
        if( im_piocheck( in, out ) )
.br
		return( -1 );
.br
        if( im_cp_desc( out, in ) ) 
.br
		return( -1 );
.br
	if( im_demand_hint( out, IM_THINSTRIP, in, NULL ) )
.br
		 return( -1 );
.br
	if( im_generate( out,
.br
		im_start_one, inv_gen, im_stop_one, in, NULL ) )
.br
		return( -1 );
.br
	return( 0 );
.br
}

.SH RETURN VALUE
All functions returns 0 on success and non-zero on error.
.SH SEE ALSO
im_generate(3), im_prepare(3). 
.SH COPYRIGHT
National Gallery
.SH AUTHOR
J. Cupitt \- 3/9/93


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/iofuncs/im_demand_hint.c
/* @(#) Hint to the evaluation mechanism that it should ask for output from
 * @(#) this image with a certain shape of patch. 
 * @(#)
 * @(#) int 
 * @(#) im_demand_hint( im, hint, in1, in2, ..., NULL )
 * @(#) IMAGE *im, *in1, *in2, ...;
 * @(#) im_demand_type hint;
 * @(#)
 * @(#) hint may be one of
 * @(#)
 * @(#)	IM_THINSTRIP
 * @(#)		This operation would like to output strips the width of the
 * @(#)		image and a few pels high. This is option suitable for
 * @(#)		point-to-point operations, such as those in the arithmetic
 * @(#)		package.
 * @(#)
 * @(#)		This is the fastest style for most simple operations.
 * @(#)
 * @(#)	IM_FATSTRIP
 * @(#)		This operation would like to output strips the width of the
 * @(#)		image and as high as possible. This option is suitable for
 * @(#)		area operations which do not violently transform coordinates,
 * @(#)		such as im_conv(). 
 * @(#)
 * @(#)	IM_SMALLTILE
 * @(#)		This is the most general demand format, and is the default.
 * @(#)		Output is demanded in small (around 100x100 pel) sections.
 * @(#)		This style works reasonably efficiently, even for bizzare
 * @(#)		operations like 45 degree rotate.
 * @(#)
 * @(#)	IM_ANY
 * @(#)		Not from a disc file, so any geometry is OK.
 * @(#)
 * @(#) NOTE: demand style falls back to the most restrictive in the pipeline.
 * @(#)	All pipeline elements in the pipeline must agree on IM_THINSTRIP
 * @(#)	before output will be asked for in this manner. If you do not set a
 * @(#) hint, you will get IM_SMALLTILE.
 * @(#)
 * @(#) in1, in2, ... are the images on which out will make demands. You
 * @(#) should terminate the list with NULL.
 * @(#)
 * @(#) int 
 * @(#) im_demand_hint_array( im, hint, in )
 * @(#) IMAGE *im, **in;
 * @(#) im_demand_type hint;
 * @(#)
 * @(#) As above, but in is a NULL-terminated array of input images. Use 
 * @(#) im_allocate_input_array() to build the input array.
 * @(#)
 * @(#) Returns non-zero on failure.
 * @(#) 
 *
 * Copyright: The National Gallery, 1993
 * Written on: 6/9/93
 * Modified on : 
 * 2/3/98 JC
 *	- IM_ANY added
 * 19/5/06
 * 	- minor change to rules: don't force ANY on no-input operations ...
 * 	  fails for image import
 * 1/12/06
 * 	- build parent/child links as well
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

/*
#define DEBUG
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

#include <vips/vips.h>
#include <vips/internal.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Max number of images we can handle.
 */
#define MAX_IMAGES (1000)

/* Given two im_demand_type, return the most restrictive.
 */
static im_demand_type
find_least( im_demand_type a, im_demand_type b )
{
	return( (im_demand_type) IM_MIN( (int) a, (int) b ) );
}

/* Set hint for this image.
 */
int 
im_demand_hint_array( IMAGE *im, im_demand_type hint, IMAGE **in )
{
	int i, len, nany;

	/* How many input images are there? And how many are IM_ANY?
	 */
	for( i = 0, len = 0, nany = 0; in[i]; i++, len++ )
		if( in[i]->dhint == IM_ANY )
			nany++;

	if( len == 0 ) 
		/* No input images? Just set the requested hint. We don't 
		 * force ANY, since the operation might be something like 
		 * tiled read of an EXR image, where we certainly don't want 
		 * ANY.
		 */
		;
	else if( nany == len ) 
		/* Special case: if all the inputs are IM_ANY, then output can 
		 * be IM_ANY regardless of what this function wants. 
		 */
		hint = IM_ANY;
	else
		/* Find the most restrictive of all the hints available to us.
		 */
		for( i = 0; i < len; i++ )
			hint = find_least( hint, in[i]->dhint );

	im->dhint = hint;

#ifdef DEBUG
        printf( "im_demand_hint_array: set dhint for \"%s\" to %s\n",
		im->filename, im_dhint2char( im->dhint ) );
#endif /*DEBUG*/

	/* im depends on all these ims.
	 */
	for( i = 0; i < len; i++ )
		im__link_make( im, in[i] );

	return( 0 );
}

/* Build an array, and call the above.
 */
int 
im_demand_hint( IMAGE *im, im_demand_type hint, ... )
{
	va_list ap;
	int i;
	IMAGE *ar[MAX_IMAGES];

	va_start( ap, hint );
	for( i = 0; i < MAX_IMAGES && (ar[i] = va_arg( ap, IMAGE * )); i++ ) 
		;
	va_end( ap );
	if( i == MAX_IMAGES ) {
		im_error( "im_demand_hint", _( "too many images" ) );
		return( -1 );
	}

	return( im_demand_hint_array( im, hint, ar ) );
}
